Bureaucracy module. Iteration 3. Roles lifecycle.

shamil-gadelshin opened the pull request:
### Features 
- roles lifecycle extrinsics
- new tests
- string errors were migrated to the Substrate `decl_error` macro errors
- better code comments

### New roles lifecycle extrinsics
- update_worker_role_account
- update_worker_reward_account
- leave_worker_role
- terminate_worker_role
- unset_lead
- unstake

Commit:
Refactor imports and fix build errors

Commit:
Improved ProposalDetails data flow

Commit:
Merge pull request #391 from Lezek123/proposal-statuses

Proposal statuses fix

Commit:
Merge pull request #394 from Lezek123/proposal-details-data-flow

Improved proposal details data flow

Commit:
Proposal withdrawal

Commit:
ActiveStake.stakeId type fix

Commit:
Merge pull request #396 from Lezek123/active-stake-type-fix

ActiveStake.stakeId type bug fix

Commit:
Merge branch 'joystream-constantinople' into francesco-better-validation

Commit:
Proposal cancellation - message

Commit:
Add Real Values for Validation and Unit To Error Messages.

Commit:
Fix Build Error

shamil-gadelshin committed:
Update proposals general parameters

shamil-gadelshin committed:
Update stake parameters for all codex proposals

shamil-gadelshin committed:
Update set_wg_mint_capacity proposal limits

Commit:
Fix units and some values

Commit:
Fix units and some values

Commit:
ReAdd Filtering Options

Commit:
Update Values and Bugfix

Commit:
Fix Build Errors

Commit:
Add Integer Check to All Fields

shamil-gadelshin committed:
Change ‘set role parameters’ proposal limits

Commit:
Bugfix

Commit:
Merge pull request #393 from Gamaranto/francesco-council-tab

Fix Council Tab crash

Commit:
Merge pull request #390 from Gamaranto/francesco-better-validation

Francesco better validation

Commit:
Refactor filterProposals and Menu

Commit:
Merge pull request #395 from Lezek123/proposal-withdrawal

Proposal withdrawal implementation

shamil-gadelshin committed:
Export proposals constants to the API metadata

Commit:
Rename Withdrawn to Canceled

Commit:
Merge pull request #398 from Gamaranto/francesco-filter-proposals

Francesco filter proposals

Commit:
Add cancellation fee fetched from runtime

shamil-gadelshin committed:
Fix tests for the ’set role parameters’ proposal

shamil-gadelshin committed:
Update ‘set validator count’ proposal limits

shamil-gadelshin committed:
Update ‘spending’ proposal limits

shamil-gadelshin committed:
Update ‘set election parameters’ proposal limits

bedeho committed:
Merge pull request #385 from shamil-gadelshin/export_proposals_consts

Export proposals constants in the API metadata

shamil-gadelshin committed:
Merge branch 'development' into update_proposals_parameters

Commit:
Minor build fix

Commit:
Extract ProfilePreview Component

Commit:
Merge branch 'joystream-constantinople' into francesco-fix-styles

Commit:
Election parameters defaults + hide irrelevant

Commit:
WG mint cap default

Commit:
Set lead default & info

Commit:
Council params form - placeholders, re-enabling limit and size fields

Commit:
Merge pull request #392 from Gamaranto/francesco-fix-styles

Francesco fix styles

Commit:
Merge pull request #399 from Lezek123/fetch-cancellation-fee

Cancellation fee: fetch from rutime

shamil-gadelshin committed:
Introduce proposal limits named constants

- introduce proposal limits named constants in the codex module

Commit:
SetLead proposal details profile preview

Lezek123 committed:
Merge branch 'joystream-constantinople' into forms-defaults

Commit:
Minor conflit resolve fix

Commit:
Merge pull request #400 from Lezek123/forms-defaults

SetElectionParameters, SetContentWorkingGroupLead, SetContentWorkingGroupMintCap - fetch current data from runtime

Commit:
Merge pull request #401 from Lezek123/set-lead-profile-preview

SetLead proposal details member profile preview

Commit:
Add MaxValidator Count Placeholders

Commit:
Add SetStorageRoleParams Placeholders and fix inconsistency in Validation

Commit:
Fix Build Errors

bedeho committed:
Merge pull request #387 from shamil-gadelshin/update_proposals_parameters

Update proposals parameters and limits.

shamil-gadelshin committed:
Update min_actors from RoleParameters

- set the value to “1” because of new requirements

Commit:
Hide not enforced parameters

Commit:
Revert Min Actors Max to 1

bedeho committed:
Merge pull request #394 from shamil-gadelshin/min_actors_parameter_fix

Update min_actors from RoleParameters

Commit:
Update the stakes

Commit:
Merge pull request #402 from Gamaranto/francesco-hide-input-fields

Hide input fields and add placeholders for SetStorageRoleParameters and SetMaxValidator count.

Commit:
Merge pull request #404 from Lezek123/hide-not-enforced

Hide not enforced parameters on proposal details page

Commit:
Merge pull request #407 from Lezek123/update-stakes

Update proposals' stakes

Commit:
Fix block hash query

Commit:
Merge pull request #408 from Lezek123/fix-block-hash-query

Fix block hash query

mnaamani committed:
increase the MaximumBlockWeight to 1_000_000_000

Commit:
Memberships - pagination

Commit:
Add voting period & fix build errors

Commit:
Set lead - allow 'NONE'

Commit:
Optimalizations

Commit:
Word wrapping for rationale and parameters

Commit:
Grace period: Expires in => Executres is

mnaamani committed:
migration: do not reset any existing rewards

Commit:
Merge pull request #421 from Lezek123/executes-in

Grace period: Expires in => Executres is

Commit:
Merge pull request #414 from Lezek123/voting-period-on-type-preview

Voting period on ProposalTypePreview

Commit:
Merge pull request #418 from Lezek123/description-wrapping

Word wrapping in proposal description and parameters

Commit:
Merge pull request #415 from Lezek123/set-empty-lead

Allow setting empty lead + members dropdown optimalizations

mnaamani committed:
Merge branch 'development' into migration-constantinople

mnaamani committed:
linter fixes

Commit:
Minor UX/UI improvements

bedeho committed:
Merge pull request #397 from mnaamani/increase-max-block-weight

Increase the MaximumBlockWeight

Commit:
Merge 'joystream-constantinople' and fix conflicts

Commit:
Build fix + minor css fix

Commit:
Details.tsx - block heights and execution error

Commit:
Add "Expired" proposals tab

Commit:
Quorum and threshold on ProposalType preview

Commit:
Build errors etc

Commit:
Remove hardcoded "tJOY" references

Commit:
lastId formula fix

Commit:
Proposal id in overview

gleb-urvanov committed:
review feedback applied

Commit:
Member loading bug fix

Commit:
Election params validation fix

Commit:
Merge pull request #426 from Lezek123/block-heights

Details.tsx - block heights and execution error

Commit:
Merge pull request #427 from Lezek123/expired-tab

Add "Expired" proposals tab

Commit:
Merge pull request #431 from Lezek123/tjoy-renaming

Remove hardcoded "tJOY" references

Commit:
Merge pull request #423 from Lezek123/ui-ux-glitches

Minor UX/UI improvements

Commit:
Merge pull request #436 from Lezek123/election-params-validation-fix

Election params validation fix

Commit:
Merge pull request #433 from Lezek123/proposal-id-overview

Proposal id in overview box

Lezek123 committed:
Merge branch 'joystream-constantinople' into quorum-thrershold

Commit:
Council - identify by membership (step 1)

Commit:
Member profile - optional id

Commit:
Minor loader fix

Commit:
Members dropdown

Commit:
Merge pull request #429 from Lezek123/quorum-thrershold

Quorum and threshold in ProposalTypePreview

deamme committed:
Remove unnecessary tabs

mnaamani committed:
runtime: make proposal discussion types ThreadId and PostId u64 like forum

Addresses https://github.com/Joystream/apps/issues/434

deamme committed:
Added ability to set custom endpoint via console

mnaamani committed:
runtime: forum make ThreadId and PostId types configurable on Trait

deamme committed:
Remove needless tabs and text

bedeho committed:
Merge pull request #442 from deamme/issue-439

Tab hell fix

mnaamani committed:
revert forum type ThreadId to u64

mnaamani committed:
clippy linter fixes

mnaamani committed:
joy-types: bump to version 0.8.0 for Constantinople release

mnaamani committed:
joy-types: remove temp file

Commit:
Fixes

mnaamani committed:
u64 ThreadId

Commit:
Merge 'joystream' and fix conflicts

Commit:
Final minor touch

mnaamani committed:
runtime: ThreadId expand comment explaining why it must be same for forum and proposal discussion

Commit:
Merge pull request #411 from Lezek123/memberships-pagination

Memberships - pagination

deamme committed:
Stop hiding information from people without a key

bedeho committed:
Merge pull request #422 from mnaamani/use-common-thread-post-id-types

Use common thread post id types

mnaamani committed:
joy-types: add missing type ThreadCounter from proposal discussion module

bedeho committed:
Merge pull request #284 from mnaamani/migration-constantinople

Runtime: Migration module - migrations for constantinople

mnaamani committed:
Merge branch 'development' into council-recurring-rewards

mnaamani committed:
Merge pull request #319 from gleb-urvanov/feature/rome-constatinople-migration-test

Feature/rome constatinople migration test

mnaamani committed:
Merge branch 'joystream-constantinople' into joy-types-bump-version

mnaamani committed:
joy-types: v0.9.0

mnaamani committed:
Merge pull request #443 from mnaamani/fix/thread-id

revert forum type ThreadId to u64

mnaamani committed:
Merge remote-tracking branch 'joystream/joystream-constantinople' into joy-types-bump-version

mnaamani committed:
Merge pull request #445 from Lezek123/joy-media-fixes-and-status-messages

Joy media fixes and status messages

mnaamani committed:
Merge branch 'joystream' into alex-rename-link-to-links-updated

deamme committed:
Fixed issues based on review

deamme committed:
Fixed issues based on second review

mnaamani committed:
travis: fix repo name

mnaamani committed:
travis: update repo secret API key

bedeho committed:
Merge pull request #275 from mnaamani/council-recurring-rewards

Runtime: Council - recurring rewards

deamme committed:
Fixed issues based on third review

Lezek123 committed:
Merge pull request #446 from deamme/issue-440

Stop hiding information from people without a key

shamil-gadelshin committed:
Add unset_lead() extrinsic to the bureacracy module

mnaamani committed:
joy-types: add missing lodash dependency and bump version to v0.9.1

Commit:
Merge pull request #444 from mnaamani/joy-types-bump-version

Bump @joystream/types package

Commit:
Merge 'joystream' and fix the conflicts

deamme committed:
Add transfer tab back

Commit:
Rename to withOnlyAccounts for consitency

shamil-gadelshin committed:
Add a test for the unset_lead() extrinsic in the bureaucracy

Lezek123 committed:
Merge pull request #454 from deamme/add-transfer-back

Add transfer tab back

shamil-gadelshin committed:
Add update_curator_role_account() extrinsic to the bureuacracy

shamil-gadelshin committed:
Add update_worker_reward_account() extrinsic with tests

shamil-gadelshin committed:
Copy leave_worker_role() extrinsic

Commit:
Change links and remove faucet references

Commit:
Merge 'joystream-constantinople' and fix conflicts

mnaamani committed:
bump runtime spec version to 13 and node binary version to v2.2.0

Constantinople

mnaamani committed:
Merge branch 'development' into constantinople-release

mnaamani committed:
runtime: reduce validator stake unbonding period to 1 day

mnaamani committed:
runtime: reward-curve adjusted for Constantinople incentive model

mnaamani committed:
runtime: reward-curve increase max_piece_count to fix precision test failure

shamil-gadelshin committed:
Add tests for leave_worker_role() extrinsic  in the bureaucracy

Commit:
Rename default WS endpoint

Commit:
Change "Tokenomics" link

Lezek123 committed:
Merge pull request #455 from Lezek123/joystream-conflicts-fix

Joystream conflicts fix

shamil-gadelshin committed:
Migrate terminate_worker_role() extrinsic

- with tests

mnaamani committed:
Merge pull request #427 from mnaamani/fix/travis-post-rename-of-repo

Fix Travis deploy scripts

mnaamani committed:
runtime: remove temp test code

Commit:
Reintroduce extrinsics, settings, storage and sudo

mnaamani committed:
build scripts: compute wasm blob hash when building the docker joystream/node image

mnaamani committed:
Merge pull request #458 from Lezek123/reintroduce-api-tabs

Reintroduce extrinsics, settings, storage and sudo

Lezek123 committed:
Merge branch 'joystream-constantinople' into references-to-older-chains

shamil-gadelshin committed:
Add unstake() extrinsic with tests to the bureaucracy module

mnaamani committed:
travis: add job to build runtime wasm blob deterministicly with docker

mnaamani committed:
travis: fix elif syntax

mnaamani committed:
Merge pull request #442 from mnaamani/travis-deploy-wasm-blob

Travis: add job to build runtime wasm blob deterministicly with docker

Commit:
Reintorduce explorer + back to light/full mode

Commit:
JavaScript, Toolbox

Lezek123 committed:
Merge pull request #457 from Lezek123/references-to-older-chains

References to older chains

shamil-gadelshin committed:
Merge pull request #434 from mnaamani/constantinople-release

Set runtime spec version to 13 and joystream-node binary version to v2.2.0

Commit:
Remove outer links, ToS and PP from side menu

mnaamani committed:
some README fixes and update runtime CHANGELOG

mnaamani committed:
readme: typo

Commit:
Membership hooks - add isSubscribed

shamil-gadelshin committed:
Merge branch 'bureaucracy' into bureaucracy_roles_lifecycle

# Conflicts:
#	runtime-modules/bureaucracy/src/lib.rs
#	runtime-modules/bureaucracy/src/tests/mod.rs

mnaamani committed:
Update README.md

Co-authored-by: Martin <35237943+bwhm@users.noreply.github.com>

mnaamani committed:
Merge pull request #444 from mnaamani/readme-changelog

README and CHANGELOG updates

mnaamani committed:
Merge branch 'master' into development

shamil-gadelshin committed:
Add comments to the bureaucracy module

shamil-gadelshin committed:
Merge pull request #445 from Joystream/development

Constantinople Release

mnaamani committed:
Merge pull request #459 from Lezek123/reintroduce-tabs-continuation

Reintorduce explorer and bring back to "light"/"full" mode

Lezek123 committed:
Merge pull request #438 from Lezek123/council-members-identity

Council - identify candidates/members by membership

bwhm committed:
Merge pull request #453 from Joystream/joystream-constantinople

Constantinople Deployment

shamil-gadelshin committed:
Introduce Substrate errors in the bureucracy module

shamil-gadelshin committed:
Migrate all errors to the decl_error macro in the bureaucracy

shamil-gadelshin committed:
Migrate other joystream module errors in the bureuacracy

shamil-gadelshin committed:
Migrate tests to the decl_macro errors in the bureaucracy

Commit:
Encoding fix + tmp rename "Links" to "Link" in UI

Commit:
Force description too

Commit:
Moving changes back to schema

bedeho commented on runtime-modules/bureaucracy/src/lib.rs:
```diff
@@ -226,10 +309,182 @@ decl_module! {
+            <CurrentLead<T, I>>::kill();
+
+            // Unwrap lead
+            if let Some(lead) = lead {
```
We statically know there is a lead, so this test cannot fail, e.g. do on L319

```Rust
let lead = ensure_lead_set()?; //Returns Ok(Lead)
```

bedeho commented on runtime-modules/bureaucracy/src/lib.rs:
```diff
@@ -226,10 +309,182 @@ decl_module! {
+        }
+
+        /// Update the associated role account of the active worker.
+        pub fn update_worker_role_account(
```
Why are there two actor identifiers here? The `worker_id` should resolve to a worker profile, which holds the underlying `member_id`, which then can be used to lookup whether account signing `origin` is same as membershp controller account.

bedeho replied:
I guess perhaps I missed this in an earlier review, but there should be `Worker.member_id`, and there would then probably need to be a `ensure_member_signed_as_worker(origin, worker_id) => Result<Worker, >`.

bedeho reviewed (changes requested):
All in all, excellent work, and very nice job switching out the error types.

- Could we use the more salient mutation safety line from before?

```Rust
           //
           // == MUTATION SAFE ==
           //
```

- It struck me that, waiting for the worker to unstake actually introduces a great deal of complexity and risk for us, and.... we don't use it for anything. It is only used to update the state from unstaking to done. Can't we just simplify and drop it all? Its worth considering at least.

mnaamani committed:
Merge pull request #461 from Lezek123/media-schemas-and-links-fix

Media schemas and links fix

mnaamani committed:
cargo: run tests and clippy in release mode

this will reduce total build times in travis, and when running git hooks

mnaamani committed:
travis and hooks, cargo test and clippy for target

mnaamani committed:
travis: cache cargo again

mnaamani committed:
travis: additional optimizations

mnaamani committed:
travis: test recompiling runtime with cargo cache enabled

mnaamani committed:
travis: fail if cargo test and clippy fail

mnaamani committed:
scripts: no need to explicitly pull a docker image

mnaamani committed:
travis: disable cache again to test build time

mnaamani committed:
travis: experiment with caching with just one travis job

mnaamani committed:
travis: re-enable cargo cache

mnaamani committed:
build.rs - make sure to trigger WASM build after dummy from clippy

Commit:
Move to subfolder

Commit:
Merge remote-tracking branch 'pioneer-upstream/monorepo-migration' into pioneer-migration

mnaamani committed:
travis: clippy before test

Commit:
Pioneer workspaces config etc.

shamil-gadelshin committed:
Add review fixes

- remove worker stages
- remove worker exit summary
- remove worker after exiting
- remove unstate()
- add member_id to the worker

mnaamani committed:
travis: use nightly for clippy to avoid messing with env variables

mnaamani committed:
travis: install clippy for nightly toolchain

bedeho reviewed (approved):
**Perfect**

mnaamani committed:
travis, build scripts use nightly

mnaamani committed:
travis: cache enabled

mnaamani committed:
travi: typo fix

mnaamani committed:
travis: rustfmt not available in nightly?

mnaamani committed:
travis: rust stable by default and add nightly

mnaamani committed:
runtime wasm builder: try newer version to ensure always building for host target

Commit:
Minor fixes / cleanup

Commit:
Pioneer webpack/tsconfig for "@joystream/types"

mnaamani committed:
travis: load cache, but dump it - just testing!

mnaamani committed:
travis: skipping clippy

mnaamani committed:
travis: -Z minimal-versions flag for cargo build

mnaamani committed:
travis: rust nightly lock to version 2020-05-22 build fails in newer version

mnaamani committed:
travis: typo

mnaamani committed:
travis: correct nightly for all steps

mnaamani committed:
travis: rust nightly 2020-05-23

mnaamani committed:
travis: too many clippy errors with nightly!

mnaamani committed:
travis: return cp task removed accidentally

mnaamani committed:
travis: try to skip wasm build when running tests

mnaamani committed:
travis: fix path to target folder

mnaamani committed:
travis: re-enable caching now that build is working

mnaamani committed:
build scripts: restore to using stable for test and clippy

mnaamani committed:
git-hooks: keep cargo test and clippy in debug mode

This way env var BUILD_DUMMY_WASM_BINARY with cargo build --release

shamil-gadelshin committed:
Merge pull request #464 from mnaamani/speedup-travis-and-hooks

Travis - Faster CI checks

mnaamani committed:
Merge branch 'development' into pioneer-migration

mnaamani committed:
Merge pull request #469 from Lezek123/pioneer-migration

Introduce Pioneer into monorepo

shamil-gadelshin committed:
Merge branch 'development' into bureaucracy_roles_lifecycle

# Conflicts:
#	runtime-modules/forum/src/lib.rs
#	runtime-modules/forum/src/mock.rs
#	runtime/src/lib.rs

shamil-gadelshin committed:
Bump the runtime version

mnaamani committed:
travis: looking for .wasm file!

mnaamani committed:
travis: remove deployment and cleanup some comments

mnaamani committed:
travis: trigger wasm build ?

mnaamani committed:
travis: double check .wasm is moved to correct location

also doing one more commit, testing env variables is correct given cache and cycling between travis jobs

mnaamani committed:
Merge branch 'travis-trigger-wasm' into travis-remove-deployment-task

shamil-gadelshin committed:
Merge pull request #501 from mnaamani/travis-remove-deployment-task

Travis: Remove github releases config

shamil-gadelshin committed:
Merge branch 'development' into bureaucracy_roles_lifecycle
